src/gui/XournalView.cpp

```diff
@@ -86,40 +95,34 @@ void XournalView::staticLayoutPages(GtkWidget* widget, GtkAllocation* allocation
     xv->layoutPages();
 }
 
+
 auto XournalView::clearMemoryTimer(XournalView* widget) -> gboolean {
-    GList* list = nullptr;
+    widget->cleanupBufferCache();
+    return true;
+}
 
-    for (auto&& page: widget->viewPages) {
-        if (page->getLastVisibleTime() > 0) {
-            list = g_list_insert_sorted(list, page, reinterpret_cast<GCompareFunc>(pageViewIncreasingClockTime));
+auto XournalView::cleanupBufferCache() -> void {
+    const auto& [pagesLower, pagesUpper] = this->preloadPageBounds(this->currentPage, this->viewPages.size());
+    g_assert(pagesLower <= pagesUpper);
+
+    // Queue pages to remove, sorted by the oldest (i.e. smallest last visible time)
+    constexpr auto cmp = [](XojPageView* a, XojPageView* b) {
+        return a->getLastVisibleTime() < b->getLastVisibleTime();
+    };
+    std::vector<XojPageView*> toRemove;
+    // Keep in cache: visible pages + preload_before + preload_after
+    for (size_t i = 0; i < this->viewPages.size(); i++) {
+        auto&& page = this->viewPages[i];
+        const size_t pageNum = i + 1;
+        const bool isPreload = pagesLower <= pageNum && pageNum <= pagesUpper;
+        if (!isPreload && page->getLastVisibleTime() > 0) {
+            toRemove.insert(std::upper_bound(toRemove.begin(), toRemove.end(), page, cmp), page);
```

Comment:
As you clear the full list, there is no need to sort the pages anymore.

Reply:
Fixed. Now it just deletes the buffers in the loop.